public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx

```diff
@@ -0,0 +1,493 @@
+  labels: SelectableLabel[];
+  searchTerm: string;
+  status: string;
+  error: string;
+  validationStatus: string;
+}
+
+interface FacettableValue {
+  name: string;
+  // facetted?: boolean; // True if value is possible given the selector
+  selected?: boolean;
+}
+
+export interface SelectableLabel {
+  name: string;
+  selected?: boolean;
+  loading?: boolean;
+  values?: FacettableValue[];
+  hidden?: boolean;
+  facets?: number;
+}
+
+export function buildSelector(labels: SelectableLabel[]): string {
+  const selectedLabels = [];
+  for (const label of labels) {
+    if (label.selected && label.values && label.values.length > 0) {
+      const selectedValues = label.values.filter((value) => value.selected).map((value) => value.name);
+      if (selectedValues.length > 1) {
+        selectedLabels.push(`${label.name}=~"${selectedValues.join('|')}"`);
+      } else if (selectedValues.length === 1) {
+        selectedLabels.push(`${label.name}="${selectedValues[0]}"`);
+      }
+    }
+  }
+  return ['{', selectedLabels.join(','), '}'].join('');
+}
+
+export function facetLabels(
+  labels: SelectableLabel[],
+  possibleLabels: Record<string, string[]>,
+  lastFacetted?: string
+): SelectableLabel[] {
+  return labels.map((label) => {
+    const possibleValues = possibleLabels[label.name];
+    if (possibleValues) {
+      let existingValues: FacettableValue[];
+      if (label.name === lastFacetted && label.values) {
+        // Facetting this label, show all values
+        existingValues = label.values;
+      } else {
+        // Keep selection in other facets
+        const selectedValues: Set<string> = new Set(
+          label.values?.filter((value) => value.selected).map((value) => value.name) || []
+        );
+        // Values for this label have not been requested yet, let's use the facetted ones as the initial values
+        existingValues = possibleValues.map((value) => ({ name: value, selected: selectedValues.has(value) }));
+      }
+      return { ...label, loading: false, values: existingValues, facets: existingValues.length };
+    }
+
+    // Label is facetted out, hide all values
+    return { ...label, loading: false, hidden: !possibleValues, values: undefined, facets: 0 };
+  });
+}
+
+const getStyles = stylesFactory((theme: GrafanaTheme) => ({
+  wrapper: css`
+    background-color: ${theme.colors.bg2};
+    padding: ${theme.spacing.md};
+    max-width: 100rem;
+  `,
+  list: css`
+    margin-top: ${theme.spacing.sm};
+    display: flex;
+    flex-wrap: wrap;
+    max-height: 200px;
+    overflow: auto;
+  `,
+  section: css`
+    & + & {
+      margin: ${theme.spacing.md} 0;
+    }
+    position: relative;
+  `,
+  selector: css`
+    font-family: ${theme.typography.fontFamily.monospace};
+    margin-bottom: ${theme.spacing.sm};
+  `,
+  status: css`
+    padding: ${theme.spacing.xs};
+    color: ${theme.colors.textSemiWeak};
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
```

Comment:
I am not sure why, but these ellipsis are causing weird border. Not sure if we can do anything about it, but noticed it.

Reply:
![image](https://user-images.githubusercontent.com/30407135/109538818-d02f4380-7ac0-11eb-9cb2-65db7dd37a01.png)